rest/oidc_api_test.go

```diff
@@ -1,139 +1,455 @@
+					IncludeAccessToken:   true,
+					DisableCallbackState: base.BoolPtr(true),
+				},
+			},
+			inputDefaultProvider: "google", wantUsername: "google_noah",
+			inputRequestURL: "/db/_oidc?provider=google&offline=true",
+		},
+		{
+			name: "callback_state_disabled_explicitly_HttpOnly_cookie",
+			inputProviders: auth.OIDCProviderMap{
+				"google": &auth.OIDCProvider{
+					Name:                        "google",
+					Issuer:                      issuerGoogle,
+					ClientID:                    "foo",
+					UserPrefix:                  "google",
+					ValidationKey:               base.StringPointer("bar"),
+					Register:                    true,
+					DiscoveryURI:                issuerGoogle + auth.DiscoveryConfigPath,
+					IncludeAccessToken:          true,
+					DisableCallbackState:        base.BoolPtr(true),
+					CallbackStateCookieHTTPOnly: true,
+				},
+			},
+			inputDefaultProvider: "google", wantUsername: "google_noah",
+			inputRequestURL: "/db/_oidc?provider=google&offline=true",
+		},
+		{
+			name: "callback_state_disabled_explicitly_HttpOnly_cookie_invalidateState",
+			inputProviders: auth.OIDCProviderMap{
+				"google": &auth.OIDCProvider{
+					Name:                        "google",
+					Issuer:                      issuerGoogle,
+					ClientID:                    "foo",
+					UserPrefix:                  "google",
+					ValidationKey:               base.StringPointer("bar"),
+					Register:                    true,
+					DiscoveryURI:                issuerGoogle + auth.DiscoveryConfigPath,
+					IncludeAccessToken:          true,
+					DisableCallbackState:        base.BoolPtr(true),
+					CallbackStateCookieHTTPOnly: true,
+				},
+			},
+			inputDefaultProvider: "google", wantUsername: "google_noah",
+			inputRequestURL: "/db/_oidc?provider=google&offline=true",
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			options := auth.OIDCOptions{Providers: tc.inputProviders, DefaultProvider: &tc.inputDefaultProvider}
+			restTesterConfig := RestTesterConfig{DatabaseConfig: &DbConfig{OIDCConfig: &options}}
+			restTester := NewRestTester(t, &restTesterConfig)
+			defer restTester.Close()
+			mockSyncGateway := httptest.NewServer(restTester.TestPublicHandler())
+			defer mockSyncGateway.Close()
+			mockSyncGatewayURL := mockSyncGateway.URL
+			invalidateState = tc.errorExpected
+			defer func() { invalidateState = false }()
+
+			// Initiate OpenID Connect Authorization Code flow.
+			requestURL := mockSyncGatewayURL + tc.inputRequestURL
+			request, err := http.NewRequest(http.MethodGet, requestURL, nil)
+			require.NoError(t, err, "Error creating new request")
+			jar, err := cookiejar.New(nil)
+			require.NoError(t, err, "Error creating new cookie jar")
+			client := &http.Client{Jar: jar}
+			response, err := client.Do(request)
+			require.NoError(t, err, "Error sending request")
+
+			if tc.errorExpected {
+				require.Equal(t, http.StatusBadRequest, response.StatusCode)
```

Comment:
Check for state mismatch error

```suggestion
				bodyBytes, err := ioutil.ReadAll(response.Body)
				require.NoError(t, err, "error reading response body")
				assert.Contains(t, string(bodyBytes), "State mismatch")
				assert.Equal(t, http.StatusBadRequest, response.StatusCode)
```

Reply:
Included check for state mismatch error